# Análise dos Mundiais de League of Legends

## Inspeção inicial dos dados

Nesse primeiro momentos, serão analisadas as dimensões, colunas, tipos e características iniciais dos datasets.

In [47]:
import pandas as pd

In [48]:
df_players = pd.read_csv("../datasets/players_stats.csv")
df_champions = pd.read_csv("../datasets/champions_stats.csv")
df_matches = pd.read_csv("../datasets/matchs_stats.csv")

## 1. Estatísticas de jogadores

O arquivo `players_stats.csv` contém características agregadas de jogadores por temporada, evento e equipe.

In [49]:
df_players.shape

(1283, 21)

In [50]:
df_players.columns

Index(['season', 'event', 'team', 'player', 'games_played', 'wins', 'loses',
       'win_rate', 'kills', 'deaths', 'assists', 'kill_death_assist_ratio',
       'creep_score', 'cs/min', 'gold', 'gold/min', 'damage', 'damage/min',
       'kill_participation', 'kill_share', 'gold_share'],
      dtype='str')

In [51]:
df_players.dtypes

season                       int64
event                          str
team                           str
player                         str
games_played                 int64
wins                         int64
loses                        int64
win_rate                   float64
kills                      float64
deaths                     float64
assists                    float64
kill_death_assist_ratio    float64
creep_score                float64
cs/min                     float64
gold                       float64
gold/min                     int64
damage                     float64
damage/min                 float64
kill_participation         float64
kill_share                 float64
gold_share                 float64
dtype: object

In [52]:
df_players.head()

,season,event,team,player,games_played,wins,loses,win_rate,kills,deaths,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
0,1,Main,Against_All_authority,Kujaa,12,7,5,58.3,0.25,2.58,...,3.32,13.58,0.34,7.8,198,NaN,NaN,NaN,NaN,NaN
1,1,Main,Against_All_authority,Linak,12,7,5,58.3,1.75,3.58,...,2.63,113.17,2.86,10.3,259,NaN,NaN,NaN,NaN,NaN
2,1,Main,Against_All_authority,MoMa,12,7,5,58.3,4.17,2.75,...,3.55,242.25,6.13,12.1,307,NaN,NaN,NaN,NaN,NaN
3,1,Main,Against_All_authority,sOAZ,12,7,5,58.3,4.00,2.92,...,3.80,214.67,5.43,11.6,293,NaN,NaN,NaN,NaN,NaN
4,1,Main,Against_All_authority,YellOwStaR,12,7,5,58.3,3.75,3.25,...,2.74,276.33,6.99,13.2,333,NaN,NaN,NaN,NaN,NaN


In [53]:
df_players.tail()

,season,event,team,player,games_played,wins,loses,win_rate,kills,deaths,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
1278,12,Main,Top_Esports,JackeyLove,6,3,3,50.0,5.33,2.33,...,4.50,295.33,9.37,15.3,486,20700.0,656.0,76.8,39.0,26.1
1279,12,Main,Top_Esports,knight,6,3,3,50.0,4.33,1.17,...,8.43,283.50,8.99,13.9,442,22400.0,711.0,72.0,31.7,23.7
1280,12,Main,Top_Esports,Mark,6,3,3,50.0,1.17,2.00,...,5.25,24.00,0.76,7.7,245,5400.0,171.0,76.8,8.5,13.1
1281,12,Main,Top_Esports,Tian,6,3,3,50.0,1.83,2.33,...,3.93,174.00,5.52,10.5,332,10900.0,344.0,67.1,13.4,17.8
1282,12,Main,Top_Esports,Wayward,6,3,3,50.0,1.00,4.00,...,1.38,254.33,8.07,11.3,360,10600.0,338.0,40.2,7.3,19.3


In [54]:
df_players.index

RangeIndex(start=0, stop=1283, step=1)

### Observações iniciais

- O conjunto possui 1283 linhas e 21 colunas.
- Cada linha parece representar estatísticas agregadas de um jogador em determinada temporada, evento e equipe.
- O DataFrame utiliza um índice numérico sequencial, iniciado em zero.
- Algumas métricas, como `damage` e `damage/min`, possuem valores ausentes nas temporadas iniciais.
- A coluna `gold/min` foi interpretada como `int64`. Por mais que seja uma taxa, os valores possivelmente foram previamente arredondados para números inteiros na fonte.

## 2. Estatísticas de champions

O arquivo `champions_stats.csv` contém características agregadas de champions por temporada e evento.

In [55]:
df_champions.shape

(1345, 24)

In [56]:
df_champions.columns

Index(['season', 'event', 'champion', 'games_contests', 'pick_ban_ratio',
       'banned_games', 'played_games', 'played_by_number_of_players', 'win',
       'lose', 'win_rate', 'kills', 'deaths', 'assists',
       'kill_death_assist_ratio', 'creep_score', 'cs/min', 'gold', 'gold/min',
       'damage', 'damage/min', 'kill_participation', 'kill_share',
       'gold_share'],
      dtype='str')

In [57]:
df_champions.dtypes

season                           int64
event                              str
champion                           str
games_contests                   int64
pick_ban_ratio                 float64
banned_games                   float64
played_games                     int64
played_by_number_of_players    float64
win                              int64
lose                             int64
win_rate                       float64
kills                          float64
deaths                         float64
assists                        float64
kill_death_assist_ratio        float64
creep_score                    float64
cs/min                         float64
gold                           float64
gold/min                       float64
damage                         float64
damage/min                     float64
kill_participation             float64
kill_share                     float64
gold_share                     float64
dtype: object

In [58]:
df_champions.head()

,season,event,champion,games_contests,pick_ban_ratio,banned_games,played_games,played_by_number_of_players,win,lose,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
0,1,Main,Nunu,27,96.4,9.0,18,6.0,12,6,...,3.63,90.17,2.46,9.4,256.0,NaN,NaN,NaN,NaN,NaN
1,1,Main,Alistar,25,89.3,12.0,13,4.0,8,5,...,4.03,114.00,3.11,9.6,261.0,NaN,NaN,NaN,NaN,NaN
2,1,Main,Rumble,23,82.1,21.0,2,2.0,2,0,...,8.33,162.00,5.02,10.1,311.0,NaN,NaN,NaN,NaN,NaN
3,1,Main,Vladimir,22,78.6,14.0,8,5.0,3,5,...,2.39,224.63,5.99,11.2,298.0,NaN,NaN,NaN,NaN,NaN
4,1,Main,Twisted Fate,22,78.6,18.0,4,2.0,2,2,...,1.67,206.50,5.43,11.4,299.0,NaN,NaN,NaN,NaN,NaN


In [59]:
df_champions.tail()

,season,event,champion,games_contests,pick_ban_ratio,banned_games,played_games,played_by_number_of_players,win,lose,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
1340,12,Main,Nasus,1,1.3,NaN,1,1.0,0,1,...,0.75,15.0,0.50,5.7,190.0,6100.0,203.0,75.0,0.0,12.1
1341,12,Main,Rumble,1,1.3,NaN,1,1.0,0,1,...,0.50,234.0,7.74,10.5,347.0,11600.0,385.0,75.0,50.0,22.1
1342,12,Main,Wukong,1,1.3,NaN,1,1.0,0,1,...,1.40,159.0,3.75,11.4,270.0,11900.0,281.0,63.6,18.2,16.9
1343,12,Main,Xin Zhao,1,1.3,NaN,1,1.0,0,1,...,1.33,179.0,6.27,9.0,316.0,7600.0,266.0,100.0,25.0,20.5
1344,12,Main,Olaf,1,1.3,1.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [60]:
df_champions.index

RangeIndex(start=0, stop=1345, step=1)

### Observações iniciais

- O conjunto possui 1345 linhas e 24 colunas.
- Cada linha parece representar estatísticas agregadas de um champion em determinada season e evento.
- O DataFrame utiliza um índice numérico sequencial, iniciado em zero.
- As colunas `banned_games` e `played_by_number_of_players` foram interpretadas como `float64`, apesar de representarem contagens. Isso provavelmente ocorre devido à presença de valores ausentes nessas colunas.
- Assim como no conjunto de dados de jogadores, algumas métricas como `damage`, `damage/min`, possuem valores ausentes nas temporadas iniciais.
- A coluna `banned_games` possui valores ausentes que podem representar campeões não banidos, mas isso ainda precisa ser verificado.

## 3. Estatísticas de partidas

O arquivo `matchs_stats.csv` informações sobre as partidas, como equipes participantes, equipe vencedora, escolhas de draft e jogadores escalados em cada posição.

In [61]:
df_matches.shape

(1070, 37)

In [62]:
df_matches.columns

Index(['season', 'date', 'event', 'patch', 'blue_team', 'red_team', 'winner',
       'ban_1_blue_team', 'ban_2_blue_team', 'ban_3_blue_team',
       'ban_4_blue_team', 'ban_5_blue_team', 'ban_1_red_team',
       'ban_2_red_team', 'ban_3_red_team', 'ban_4_red_team', 'ban_5_red_team',
       'pick_1_blue_team', 'pick_2_blue_team', 'pick_3_blue_team',
       'pick_4_blue_team', 'pick_5_blue_team', 'pick_1_red_team',
       'pick_2_red_team', 'pick_3_red_team', 'pick_4_red_team',
       'pick_5_red_team', 'top_blue_team', 'jungler_blue_team',
       'mid_blue_team', 'adc_blue_team', 'support_blue_team', 'top_red_team',
       'jungler_red_team', 'mid_red_team', 'adc_red_team', 'support_red_team'],
      dtype='str')

In [63]:
df_matches.dtypes

season                 int64
date                     str
event                    str
patch                float64
blue_team                str
red_team                 str
winner                   str
ban_1_blue_team          str
ban_2_blue_team          str
ban_3_blue_team          str
ban_4_blue_team          str
ban_5_blue_team          str
ban_1_red_team           str
ban_2_red_team           str
ban_3_red_team           str
ban_4_red_team           str
ban_5_red_team           str
pick_1_blue_team         str
pick_2_blue_team         str
pick_3_blue_team         str
pick_4_blue_team         str
pick_5_blue_team         str
pick_1_red_team          str
pick_2_red_team          str
pick_3_red_team          str
pick_4_red_team          str
pick_5_red_team          str
top_blue_team            str
jungler_blue_team        str
mid_blue_team            str
adc_blue_team            str
support_blue_team        str
top_red_team             str
jungler_red_team         str
mid_red_team  

In [64]:
df_matches.head()

,season,date,event,patch,blue_team,red_team,winner,ban_1_blue_team,ban_2_blue_team,ban_3_blue_team,...,top_blue_team,jungler_blue_team,mid_blue_team,adc_blue_team,support_blue_team,top_red_team,jungler_red_team,mid_red_team,adc_red_team,support_red_team
0,1,2011-06-20,Main,NaN,Against_All_authority,Fnatic,Fnatic,Alistar,Gragas,Anivia,...,MoMa,Linak,sOAZ,YellOwStaR,Kujaa,Shushei,Cyanide,xPeke,Lamia,Mellisan
1,1,2011-06-20,Main,NaN,Fnatic,Against_All_authority,Against_All_authority,Nunu,Warwick,Amumu,...,Shushei,Cyanide,xPeke,Lamia,Mellisan,sOAZ,Linak,MoMa,YellOwStaR,Kujaa
2,1,2011-06-20,Main,NaN,Against_All_authority,TSM,Against_All_authority,Jarvan IV,Annie,Singed,...,YellOwStaR,Linak,MoMa,sOAZ,Kujaa,Chaox,TheOddOne,Reginald,TheRainMan,Xpecial
3,1,2011-06-20,Main,NaN,Epik_Gamer,TSM,TSM,Zilean,Vayne,Alistar,...,Dyrus,Dan Dinh,Salce,Westrice,Doublelift,TheRainMan,TheOddOne,Reginald,Chaox,Xpecial
4,1,2011-06-20,Main,NaN,TSM,Epik_Gamer,TSM,Akali,Vladimir,Taric,...,TheRainMan,TheOddOne,Reginald,Chaox,Xpecial,Westrice,Dan Dinh,Salce,Dyrus,Doublelift


In [65]:
df_matches.tail()

,season,date,event,patch,blue_team,red_team,winner,ban_1_blue_team,ban_2_blue_team,ban_3_blue_team,...,top_blue_team,jungler_blue_team,mid_blue_team,adc_blue_team,support_blue_team,top_red_team,jungler_red_team,mid_red_team,adc_red_team,support_red_team
1065,12,2022-10-08,Main,12.18,T1,EDward_Gaming,T1,LeBlanc,Maokai,Graves,...,Zeus,Oner,Faker,Gumayusi,Keria,Flandre,JieJie,Scout,Viper,Meiko
1066,12,2022-10-08,Main,12.18,JD_Gaming,Evil_Geniuses_2e-NA,JD_Gaming,Maokai,Akali,Ornn,...,369,Kanavi,Yagao,Hope,Missing,Impact,Inspired,jojopyun,Kaori,Vulcan
1067,12,2022-10-07,Main,12.18,CTBC_Flying_Oyster,100_Thieves,CTBC_Flying_Oyster,Yuumi,Sylas,Seraphine,...,Rest,Gemini,Mission,Shunn,Koala,Ssumday,Closer,Abbedagge,FBI,huhi
1068,12,2022-10-07,Main,12.18,G2_Esports,DWG_KIA,DWG_KIA,Aatrox,Maokai,Varus,...,BrokenBlade,Jankos,caPs,Flakked,Targamas,Nuguri,Canyon,ShowMaker,deokdam,Kellin
1069,12,2022-10-07,Main,12.18,Cloud9,Fnatic,Fnatic,Graves,Trundle,Caitlyn,...,Fudge,Blaber,Jensen,Berserker,Zven,Wunder,Razork,Humanoid,Upset,Hylissang


In [66]:
df_matches.index

RangeIndex(start=0, stop=1070, step=1)

### Observações iniciais

- O conjunto possui 1070 linhas e 37 colunas.
- Cada linha parece representar uma partida, incluindo seu resultado, equipes participantes, detalhes de draft e jogadores escalados em cada posição.
- O DataFrame utiliza um índice numérico sequencial, iniciado em zero.
- A coluna `date` foi interpretada como `str`, ainda não está representada por um tipo específico de data.
- A coluna `patch` possui valores ausentes em temporadas iniciais.

## 4. Verificação da qualidade dos dados

Nesta etapa, serão investigados valores ausentes, registros duplicados e possíveis inconsistências nos conjuntos de dados. Nenhuma alteração será realizada antes de compreender o significado desses problemas.

### 4.1 Qualidade dos dados de jogadores

In [67]:
df_players.duplicated().sum()

np.int64(0)

In [68]:
missing_values_players = df_players.isna().sum()

missing_values_players[
    missing_values_players > 0
]

damage                874
damage/min            874
kill_participation    172
kill_share            172
gold_share            172
dtype: int64

In [69]:
missing_values_players_percentage = missing_values_players / df_players.shape[0] * 100

missing_values_players_percentage[
    missing_values_players_percentage > 0
]

damage                68.12159
damage/min            68.12159
kill_participation    13.40608
kill_share            13.40608
gold_share            13.40608
dtype: float64

#### Observações sobre a qualidade dos dados

- O conjunto não possui linhas completamente duplicadas.
- As colunas `damage`, `damage/min`, `kill_participation`, `kill_share` e `gold_share` possuem valores ausentes.
- As colunas com maior quantidade de valores ausentes são `damage` e `damage/min`, ambas com aproximadamente 68% dos seus valores ausentes.
- Inicialmente, os valores ausentes parecem estar concentrados em métricas que não eram registradas nas primeiras temporadas.

#### Investigação dos valores ausentes por temporada

A inspeção inicial sugere que os valores ausentes podem estar concentrados nas primeiras temporadas, quando algumas métricas ainda não eram registradas.

Para investigar essa hipótese, será comparada, em cada temporada, a quantidade total de registros com a quantidade de valores preenchidos nas colunas `damage` e `kill_participation`.

In [70]:
players_by_season = df_players.groupby("season")["player"].count()

players_by_season

season
1      41
2      60
3      71
4      82
5      85
6      86
7     149
8     153
9     147
10    131
11    133
12    145
Name: player, dtype: int64

A seguir, será contabilizada a quantidade de valores não ausentes em `damage` para cada temporada.

In [71]:
damage_values_by_season = df_players.groupby("season")["damage"].count()

damage_values_by_season

season
1       0
2       0
3       0
4       0
5       0
6       0
7       0
8       0
9       0
10    131
11    133
12    145
Name: damage, dtype: int64

In [72]:
kill_participation_by_season = df_players.groupby("season")["kill_participation"].count()

kill_participation_by_season

season
1       0
2       0
3       0
4      82
5      85
6      86
7     149
8     153
9     147
10    131
11    133
12    145
Name: kill_participation, dtype: int64

#### Conclusão da investigação

- A métrica `damage` só possui valores preenchidos a partir da temporada 10. 
- Nas temporadas em que `damage` é registrado, não existem valores ausentes nessa coluna.
- A métrica `kill_participation` só possui valores preenchidos a partid da temporada 4.
- Assim como ocorre em `damage`, a métrica `kill_participation` não possui valores ausentes nas temporadas em que essa métrica é registrada.  

Portanto, as ausências observadas não parecem ocorrer de maneira aleatória. Elas realmente estão associadas às temporadas em que essas métricas não eram registradas no conjunto de dados.

#### Verificação das demais métricas ausentes

Após identificar o padrão temporal de `damage` e `kill_participation`, será verificado se as demais colunas com valores ausentes seguem o mesmo comportamento.

In [73]:
missing_metrics_by_season = df_players.groupby("season")[
    [
        "damage",
        "damage/min",
        "kill_participation",
        "kill_share",
        "gold_share"
    ]
].count()

missing_metrics_by_season

,damage,damage/min,kill_participation,kill_share,gold_share
season,,,,,
1,0,0,0,0,0
2,0,0,0,0,0
3,0,0,0,0,0
4,0,0,82,82,82
5,0,0,85,85,85
6,0,0,86,86,86
7,0,0,149,149,149
8,0,0,153,153,153
9,0,0,147,147,147


- As métricas `damage` e `damage/min` apresentam o mesmo padrão de preenchimento e só são registradas a partir da temporada 10.
- As métricas `kill_participation`, `kill_share` e `gold_share` apresentam o mesmo padrão e começam a ser registradas na temporada 4.
- Não existem ausências pontuais nas temporadas em que essas métricas são disponibilizadas.

#### Síntese da qualidade dos dados de jogadores

O conjunto não possui registros completamente duplicados. Os valores ausentes seguem um padrão temporal associado à disponibilidade das métricas em diferentes temporadas, não havendo sinais, até o momento, de ausências pontuais nas temporadas em que cada métrica é registrada.

### 4.2 Qualidade dos dados de campeões

In [74]:
df_champions.duplicated().sum()

np.int64(0)

In [75]:
missing_values_champions = df_champions.isna().sum()

missing_values_champions[
    missing_values_champions > 0
]

banned_games                   379
played_by_number_of_players     52
win_rate                        52
kills                           52
deaths                          52
assists                         52
kill_death_assist_ratio         52
creep_score                     52
cs/min                          52
gold                            52
gold/min                        52
damage                         865
damage/min                     865
kill_participation             222
kill_share                     222
gold_share                     220
dtype: int64

In [76]:
missing_values_champions_percentage = missing_values_champions / df_champions.shape[0] * 100

missing_values_champions_percentage[
    missing_values_champions_percentage > 0
]

banned_games                   28.178439
played_by_number_of_players     3.866171
win_rate                        3.866171
kills                           3.866171
deaths                          3.866171
assists                         3.866171
kill_death_assist_ratio         3.866171
creep_score                     3.866171
cs/min                          3.866171
gold                            3.866171
gold/min                        3.866171
damage                         64.312268
damage/min                     64.312268
kill_participation             16.505576
kill_share                     16.505576
gold_share                     16.356877
dtype: float64

#### Observações sobre a a qualidade dos dados

- O conjunto não possui linhas completamente duplicadas.
- As colunas com maior quantidade de valores ausentes são `damage` e `damage/min`, ambas com aproximadamente 64% dos seus valores ausentes.
- Parte dos valores ausentes parece estar relacionada a métricas que ainda não eram registradas nas primeiras temporadas.
- Também é possível observar que, quando um campeão não possui partidas disputadas em determinada temporada e evento, suas estatísticas de desempenho aparecem como valores ausentes.

#### Investigação de valores ausentes causados por campeões não utilizados

A inspeção inicial sugere que os valores ausentes podem ter causas diferentes dentro do conjunto de dados de campeões.

Para investigar essa hipótese, serão analisadas as linhas em que a coluna `played_games` é igual a zero, a fim de verificar o impacto disso nas outras colunas.

In [77]:
champions_without_games = df_champions[
    df_champions['played_games'] == 0
]

champions_without_games.head()

,season,event,champion,games_contests,pick_ban_ratio,banned_games,played_games,played_by_number_of_players,win,lose,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
71,2,Main,Twisted Fate,7,22.6,7.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
89,2,Main,Sivir,3,9.7,3.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
96,2,Main,Draven,2,6.5,2.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
103,2,Main,Gangplank,1,3.2,1.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
104,2,Main,Twitch,1,3.2,1.0,0,NaN,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [78]:
champions_without_games.shape

(52, 24)

In [79]:
champions_without_games.isna().sum()

season                          0
event                           0
champion                        0
games_contests                  0
pick_ban_ratio                  0
banned_games                    0
played_games                    0
played_by_number_of_players    52
win                             0
lose                            0
win_rate                       52
kills                          52
deaths                         52
assists                        52
kill_death_assist_ratio        52
creep_score                    52
cs/min                         52
gold                           52
gold/min                       52
damage                         52
damage/min                     52
kill_participation             52
kill_share                     52
gold_share                     52
dtype: int64

##### Conclusão sobre os campeões sem partidas

- 52 registros possuem `played_games` igual a zero.
- Em todos esses registros as seguintes métricas ficam ausentes: `played_by_number_of_players`, `win_rate`, `kills`, `deaths`, `assists`, `kill_death_assist_ratio`, `creep_score`, `cs/min`, `gold`, `gold/min`, `damage`, `damage/min`, `kill_participation`, `kill_share` e `gold_share`.
- Mesmo quando o campeão não é utilizado as seguintes métricas permanecem preenchidas: `season`, `event`, `champion`, `games_contests`, `pick_ban_ratio`, `banned_games`, `win` e `lose`
- Essa investigação explica parte das ausências do dataset, mas não explica completamente. Nos 52 registros em que `played_games` é igual a zero, `banned_games` não possui valores ausentes. Porém, analisando o conjunto como um todo, 28% dos valores de `banned_games` estão ausentes. 
- Outra parte dos valores ausentes está relacionada às temporadas iniciais, em que algumas métricas como `damage` e `damage/min` ainda não eram registradas.

#### Investigação dos valores ausentes em `banned_games`

A coluna `banned_games` possui aproximadamente 28% dos valores ausentes. Será investigado se essas ausências representam campeões que não foram banidos em nenhuma partida ou se indicam falta de informação. 

In [80]:
champions_without_ban_data = df_champions[
    df_champions["banned_games"].isna()
]

champions_without_ban_data.head()

,season,event,champion,games_contests,pick_ban_ratio,banned_games,played_games,played_by_number_of_players,win,lose,...,kill_death_assist_ratio,creep_score,cs/min,gold,gold/min,damage,damage/min,kill_participation,kill_share,gold_share
13,1,Main,Malzahar,16,57.1,NaN,16,6.0,9,7,...,3.81,224.44,6.12,11.4,310.0,NaN,NaN,NaN,NaN,NaN
19,1,Main,Brand,8,28.6,NaN,8,3.0,3,5,...,2.82,213.25,5.41,10.7,272.0,NaN,NaN,NaN,NaN,NaN
20,1,Main,Miss Fortune,8,28.6,NaN,8,4.0,2,6,...,1.87,249.38,6.64,11.3,302.0,NaN,NaN,NaN,NaN,NaN
28,1,Main,Swain,5,17.9,NaN,5,4.0,2,3,...,2.40,186.20,5.29,9.3,266.0,NaN,NaN,NaN,NaN,NaN
29,1,Main,Caitlyn,4,14.3,NaN,4,3.0,0,4,...,0.47,220.75,6.96,8.3,261.0,NaN,NaN,NaN,NaN,NaN


In [81]:
champions_without_ban_data[
    [
        "season",
        "event",
        "champion",
        "games_contests",
        "banned_games",
        "played_games"
    ]
].head(20)

,season,event,champion,games_contests,banned_games,played_games
13,1,Main,Malzahar,16,NaN,16
19,1,Main,Brand,8,NaN,8
20,1,Main,Miss Fortune,8,NaN,8
28,1,Main,Swain,5,NaN,5
29,1,Main,Caitlyn,4,NaN,4
30,1,Main,Cho'Gath,3,NaN,3
31,1,Main,Tristana,3,NaN,3
32,1,Main,Trundle,3,NaN,3
34,1,Main,Lee Sin,2,NaN,2
35,1,Main,Teemo,2,NaN,2


In [82]:
(champions_without_ban_data["games_contests"] == champions_without_ban_data["played_games"]).value_counts()

True    379
Name: count, dtype: int64

##### Conclusão sobre valores ausentes em `banned_games`

- Em todos os registros em que `banned_games` está ausente, o valor de `games_contests` é igual ao de `played_games`. Dessa forma, nesses casos, todas as aparições do campeão no draft correspondem a partidas jogadas, não havendo registros de banimento.
- Sendo assim, há evidências consistentes de que os valores ausentes em `banned_games` representam zero banimentos e podem ser substituídos por 0 para fins de análise.

#### Investigação das métricas ausentes por temporada

Tendo identificado as ausências relacionadas a campeões sem partidas e os casos em que `banned_games` representa zero banimentos, será investigada a disponibilidade histórica das outras métricas.

Para garantir que campeões sem partidas não interfiram nessa análise, serão considerados somente os registros em que `played_games` é maior que zero.

In [83]:
champions_with_games = df_champions[
    df_champions["played_games"] > 0
]

champions_with_games.shape

(1293, 24)

In [84]:
champions_with_games_by_season = champions_with_games.groupby("season")["champion"].count()

champions_with_games_by_season

season
1      46
2      54
3      68
4      59
5      74
6      56
7     138
8     150
9     168
10    149
11    153
12    178
Name: champion, dtype: int64

In [85]:
champion_metrics_by_season = champions_with_games.groupby("season")[
    [
        "damage",
        "damage/min",
        "kill_participation",
        "kill_share",
        "gold_share"
    ]
].count()

champion_metrics_by_season

,damage,damage/min,kill_participation,kill_share,gold_share
season,,,,,
1,0,0,0,0,0
2,0,0,0,0,0
3,0,0,0,0,0
4,0,0,59,59,59
5,0,0,74,74,74
6,0,0,56,56,56
7,0,0,137,137,138
8,0,0,149,149,150
9,0,0,168,168,168


##### Conclusão sobre a disponibilidade histórica das métricas

- As métricas `damage` e `damage/min` começam a ser registradas a partir da temporada 10.
- As métricas `kill_participation`, `kill_share` e `gold_share` começam a ser registradas a partir da temporada 4.
- Existe preenchimento parcial das métricas `kill_participation`, `kill_share` e `gold_share` nas temporadas 7 e 8. Logo, a dispobinilidade dessas métricas nessas temporadas precisa ser investigada com mais detalhes.

#### Investigação do preenchimento parcial nas temporadas 7 e 8

Apesar de `kill_participation`, `kill_share` e `gold_share` começarem a ser registradas na temporada 4, foi identificado preenchimento parcial nas temporadas 7 e 8.

Será investigado se essas ausências estão relacionadas às diferentes etapas do campeonato representadas pela coluna `event`.

In [86]:
champions_season_7_and_8 = champions_with_games[
    champions_with_games['season'].isin([7,8])
]

champions_season_7_and_8.shape

(288, 24)

In [87]:
champions_season_7_and_8.groupby(
    ["season", "event"]
    )["champion"].count()

season  event  
7       Main       75
        Play in    63
8       Main       74
        Play in    76
Name: champion, dtype: int64

In [88]:
champions_season_7_and_8.groupby(
    ["season", "event"]
)[
    [
        "kill_participation",
        "kill_share",
        "gold_share"
    ]
].count()

kill_participation  kill_share  gold_share
season event                                              
7      Main                     74          74          75
       Play in                  63          63          63
8      Main                     74          74          74
       Play in                  75          75          76

In [89]:
champions_season_7_and_8[
    champions_season_7_and_8[
        [
            "kill_participation",
            "kill_share",
            "gold_share"
        ]
    ].isna().any(axis=1)
][
    [
        "season",
        "event",
        "champion",
        "played_games",
        "kill_participation",
        "kill_share",
        "gold_share"
    ]
]

,season,event,champion,played_games,kill_participation,kill_share,gold_share
504,7,Main,Jinx,1,NaN,NaN,25.4
585,8,Play in,Nidalee,1,NaN,NaN,19.7


##### Conclusão sobre o preenchimento parcial

- Nas temporadas 7 e 8, as métricas `kill_participation` e `kill_share` apresentam apenas duas ausências pontuais.
- Na temporada 7, os valores ausentes pertencem ao registro da campeã Jinx no evento principal.
- Na temporada 8, os valores ausentes pertencem ao registro da campeã Nidalee no Play-In.
- Em ambos os casos, o campeão foi utilizado em apenas uma partida, mas a métrica `gold_share` permanece preenchida.
- Portanto, as ausências não parecem estar relacionadas às diferentes etapas do campeonato, já que o restante dos registros dos mesmos eventos estão preenchidos.
- Esses dois casos podem representar falhas pontuais de registro ou indisponibilidade das métricas na fonte original.

##### Síntese da qualidade dos dados de campeões

O conjunto não possui linhas completamente duplicadas. Os valores ausentes apresentam diferentes significados:

- Em `banned_games`, as ausências representam zero banimentos.
- Nos registros em que `played_games` é igual a zero, as estatísticas de desempenho não se aplicam porque o campeão não foi utilizado.
- Algumas métricas não eram registradas nas temporadas iniciais: `kill_participation`, `kill_share` e `gold_share` começam a ser disponibilizadas na temporada 4, enquanto `damage` e `damage/min` começam na temporada 10.
- Foram identificadas duas ausências pontuais em `kill_participation` e `kill_share`: Jinx na temporada 7 e Nidalee na temporada 8. A causa desses casos não pode ser determinada apenas com os dados disponíveis.

### 4.3 Qualidade dos dados de partidas